### PR TITLE
Fix self signed certificate error when connecting to Mongo emulator

### PR DESCRIPTION
**Pull request by wwlorey**
Fixes https://github.com/microsoft/vscode-cosmosdb/issues/1241

### CODE REVIEW

File: src/mongo/connectToMongoClient.ts [hunk 1]
```diff
@@ -15,6 +15,11 @@ export async function connectToMongoClient(connectionString: string, appName: st
         useNewUrlParser: true
     };
 
+    if (!!connectionString.match(/^mongodb:\/\/(localhost|127\.0\.0\.1)/)) {
+        // Prevents self signed certificate error https://github.com/microsoft/vscode-cosmosdb/issues/1241#issuecomment-614446198
+        options.tlsAllowInvalidCertificates = true;
```
Comment (EricJizbaMSFT) [inline src/mongo/connectToMongoClient.ts:20]: I'm probably fine with this, but I'd like to see a more detailed explanation when you're loosening security-related settings. Basically, convince me that you're not increasing risk for our users

Comment (wwlorey) [reply src/mongo/connectToMongoClient.ts:20]: Mongo recommends that valid certificates should definitely be used in production environments ([source](https://docs.mongodb.com/manual/tutorial/configure-ssl/)). But since the emulator is likely being used for development, it seems fine to allow invalid certificates here. 

Would it be better to throw up a warning before setting `tlsAllowInvalidCertificates`? Just so this isn't happening entirely under the covers

Comment (EricJizbaMSFT) [reply src/mongo/connectToMongoClient.ts:20]: Well why do invalid certificate errors happen? I feel like I've never run into this error and I've used a local mongo instance - can you repro?

Comment (wwlorey) [reply src/mongo/connectToMongoClient.ts:20]: Yeah I was able to repro along with Nathan & CTI. It looks like certificate behavior changed in mongo v4.0: https://docs.mongodb.com/manual/release-notes/4.0/#x-509-authentication-certificate-restrictions. So it makes sense that we're just now seeing this type of error

File: src/mongo/connectToMongoClient.ts [hunk 2]
```diff
@@ -15,6 +15,11 @@ export async function connectToMongoClient(connectionString: string, appName: st
         useNewUrlParser: true
     };
 
+    if (!!connectionString.match(/^mongodb:\/\/(localhost|127\.0\.0\.1)/)) {
```
Comment (EricJizbaMSFT) [inline src/mongo/connectToMongoClient.ts:18]: nit: I prefer `.test` over `.match` and I almost always include the `i` flag to be safe
```suggestion
    if (/^mongodb:\/\/(localhost|127\.0\.0\.1)/i.test(connectionString)) {
```

Comment (EricJizbaMSFT) [inline src/mongo/connectToMongoClient.ts:18]: [this](https://github.com/microsoft/vscode-cosmosdb/blob/master/test/mongoConnectionStrings.test.ts#L50) seems to imply that "localhost" isn't always the first thing after "mongodb://"

Might be worth creating an `isLocalhost` function and adding unit tests against the same connection strings in those tests. Never hurts to be careful when dealing with security stuff

File: src/mongo/connectToMongoClient.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply src/mongo/connectToMongoClient.ts:20]: Are you reproducing with the Cosmos DB emulator for mongo or just plain old mongodb? If it only repros in the Cosmos DB case, we might not want to do this for everyone

Comment (wwlorey) [reply src/mongo/connectToMongoClient.ts:20]: I've only ever seen the error when using the emulator. Are you suggesting we only set this flag if the connection string is for the Cosmos DB emulator?

Comment (EricJizbaMSFT) [reply src/mongo/connectToMongoClient.ts:20]: Yeah

File: main.js [hunk 3]
```diff
@@ -16,10 +16,10 @@ let perfStats = {
 
 Object.defineProperty(exports, "__esModule", { value: true });
 
-const extension = require('./out/src/extension');
+const extension = require('./dist/extension.bundle');
```
Comment (EricJizbaMSFT) [inline main.js:19]: revert these changes

File: src/mongo/connectToMongoClient.ts [hunk 4]
```diff
@@ -37,3 +42,8 @@ export class MongoConnectError extends Error {
         super(`Unable to connect to local Mongo DB instance. Make sure it is started correctly. See ${Links.LocalConnectionDebuggingTips} for tips.`);
     }
 }
+
+export function isLocalEmulatorConnectionString(connectionString: string): boolean {
+    const localEmulatorRegex: RegExp = new RegExp(`(localhost|127\.0\.0\.1):${encodeURIComponent(emulatorPassword)}`, 'i');
```
Comment (EricJizbaMSFT) [inline src/mongo/connectToMongoClient.ts:47]: What do you think - could we just check `return connectionString.includes(encodeURIComponent(emulatorPassword));`? Maybe check for the un-encoded password, too (is that possible?)

Seems virtually impossible that exact string would show up in other connection strings.

File: src/mongo/connectToMongoClient.ts [hunk 5]
```diff
@@ -37,3 +42,8 @@ export class MongoConnectError extends Error {
         super(`Unable to connect to local Mongo DB instance. Make sure it is started correctly. See ${Links.LocalConnectionDebuggingTips} for tips.`);
     }
 }
+
+export function isLocalEmulatorConnectionString(connectionString: string): boolean {
```
Comment (EricJizbaMSFT) [inline src/mongo/connectToMongoClient.ts:46]: Can you rename this to `isCosmosEmulatorConnectionString`

File: src/mongo/connectToMongoClient.ts (hunk 4 above)
Comment (wwlorey) [reply src/mongo/connectToMongoClient.ts:47]: Yeah that makes sense. It seems unlikely that `emulatorPassword` would be used elsewhere.

I tried attaching the emulator with the un-encoded password but the extension threw an error:

Screen Shot 2020-07-09 at 12 33 21 PM

So I'll just verify the connection string includes the encoded password.
